### PR TITLE
BUG: #10645 in using MultiIndex.__contains__

### PR DESCRIPTION
**Pull request by scari**
This PR fix a BUG #10645 
@sinhrks would you review my PR?

### CODE REVIEW

File: pandas/index.pyx [hunk 1]
```diff
@@ -143,7 +143,10 @@ cdef class IndexEngine:
                 return self._get_loc_duplicates(val)
             values = self._get_index_values()
             loc = _bin_search(values, val) # .searchsorted(val, side='left')
-            if util.get_value_at(values, loc) != val:
+            try:
+                if util.get_value_at(values, loc) != val:
+                    raise KeyError(val)
```
Comment (jreback) [inline pandas/index.pyx]: this just papers over things and does not solve the issue

**Comment by jreback** [comment]
needs tests

**Comment by scari** [comment]
Thanks for mentioning about missing test. Just added. :)

**Comment by scari** [comment]
I revisit this issue and Yes, you're right @jreback I did further investigation and it seems a odd behavior of Cython. so this PR doesn't solve the issue.

I'd like to finish this issue. Do you have any idea with this? Anything would be appreciated.

**Comment by jreback** [comment]
so the indexes have different behavior if > 1M elements on lookups. so that is a different path that is taken.

**Comment by scari** [comment]
This isn't Cython issue.

https://github.com/pydata/pandas/blob/master/pandas/index.pyx#L141-L148
If > 1M elements on lookups, `IndexEngine.get_loc()` do `_bin_search()` and try to `get_value_at()` the `loc`.
If the `loc` is equal to `len(values)` (like this error case) it looks ok to raise `IndexError` in `util.get_value_at()` because it means the `loc` is out of bounds.

So we should handle this `IndexError` somewhere.
IMHO, `get_loc()` itself has a 'key' parameter and if it fails key lookup, it's okay to raise `KeyError` here.

How's your thought? @jreback

**Comment by scari** [comment]
This PR also fixes #10692 
@jreback Any comments would be appreciated.

**Comment by sinhrks** [comment]
@ scari Gr8! Could you add test cases and release note for #10692? Then, pls squash to a single commit.

**Comment by jreback** [comment]
this is not correct 
you are fixing a symptom not the problem
I don't think the indexer is setup correctly

**Comment by scari** [comment]
Okay. I'll take another look at this.

**Comment by kawochen** [comment]
@scari I'll submit a PR if you don't have time to come back to this.

**Comment by jreback** [comment]
replaced by #11049

**Comment by scari** [comment]
@kawochen Thanks!

**Pull request closed without merging**